pkg/controller/master/image/controller.go

```diff
@@ -2,22 +2,231 @@ package image
+	return image, util.MinioClient.RemoveObject(util.BucketName, image.Name)
+}
+
+func (h *Handler) importImageToMinio(ctx context.Context, cancel context.CancelFunc, image *apisv1alpha1.Image) error {
+	logrus.Debugln("start importing image to minio")
+	client := http.Client{
+		//timeout by calling cancel when idle time is exceeded
+	}
+	rr, err := http.NewRequest("GET", image.Spec.URL, nil)
+	if err != nil {
+		return err
+	}
+	rr = rr.WithContext(ctx)
+
+	resp, err := client.Do(rr)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("expected 200 status from %s, got %s", image.Spec.URL, resp.Status)
+	}
+	size, err := strconv.Atoi(resp.Header.Get("Content-Length"))
+	if err != nil {
+		return err
+	}
+	fileSize := int64(size)
+	contentType := resp.Header.Get("Content-Type")
+	reader := &util.CountingReader{
+		Reader: resp.Body,
+		Total:  fileSize,
+	}
+
+	toUpdate := image.DeepCopy()
+	apisv1alpha1.ImageImported.Unknown(toUpdate)
+	apisv1alpha1.ImageImported.Message(toUpdate, "started image importing")
+	toUpdate.Status.AppliedURL = toUpdate.Spec.URL
+	toUpdate.Status.Progress = 0
+	toUpdate, err = h.UpdateStatusRetryOnConflict(toUpdate)
+	if err != nil {
+		return err
+	}
+
+	go h.syncProgress(ctx, cancel, reader, image)
+
+	n, err := util.MinioClient.PutObjectWithContext(ctx, util.BucketName, image.Name, reader, fileSize, minio.PutObjectOptions{ContentType: contentType})
+	if err != nil {
+		return err
+	}
+
+	//Succeed
+	currentImage, err := h.imageCache.Get(image.Namespace, image.Name)
+	if err != nil {
+		return err
+	}
+	toUpdate = currentImage.DeepCopy()
+	toUpdate.Status.Progress = 100
+	apisv1alpha1.ImageImported.True(toUpdate)
+	apisv1alpha1.ImageImported.Message(toUpdate, "completed image importing")
+	toUpdate.Status.DownloadURL = fmt.Sprintf("%s/%s/%s", config.MinioURL, util.BucketName, image.Name)
+	if _, err := h.UpdateStatusRetryOnConflict(toUpdate); err != nil {
+		return err
+	}
+	logrus.Debugf("Successfully imported image in size %d from %s\n", n, image.Spec.URL)
 	return nil
 }
 
-type imageController struct {
+func (h *Handler) UpdateStatusRetryOnConflict(image *apisv1alpha1.Image) (*apisv1alpha1.Image, error) {
+	retry := 3
+	for i := 0; i < retry; i++ {
+		current, err := h.imageCache.Get(image.Namespace, image.Name)
+		if err != nil {
+			return nil, err
+		}
+		if current.DeletionTimestamp != nil {
+			return current, nil
+		}
+		toUpdate := current.DeepCopy()
+		toUpdate.Status = image.Status
+		current, err = h.images.Update(toUpdate)
+		if err == nil {
+			return current, nil
+		}
+		if !apierrors.IsConflict(err) {
+			return nil, err
+		}
+		time.Sleep(2 * time.Second)
+	}
+	return nil, errors.New("failed to update status, max retries exceeded")
+}
+
+func resetImageStatus(image *apisv1alpha1.Image) {
+	image.Status.AppliedURL = ""
+	image.Status.Conditions = nil
+	image.Status.Progress = 0
+	image.Status.DownloadURL = ""
```

Comment:
Should we only reset the specific condition `ImageImported` instead of setting nil to all conditions?